implement Question Domain

daehwan2da opened the pull request:
## ✒️ 관련 이슈번호

- Closes #22

## 🔑 Key Changes

- 오늘의 질문 조회 ([notion link](https://www.notion.so/API-bf40aa0547dc4e0b8a3b4ad1ed92e1f9?p=88336bcea7e94ce3b625f76dc5199b0c&pm=s))
- 내가 답변한 질문 월별 조회 ([notion link](https://www.notion.so/API-bf40aa0547dc4e0b8a3b4ad1ed92e1f9?p=c59b9be3fff149ddb49beb669d844c68&pm=s))
- 답변 기록이 있는 연월 얻기 ([notion link](https://www.notion.so/API-bf40aa0547dc4e0b8a3b4ad1ed92e1f9?p=6ed200bdbb61466786db5947cef8664e&pm=s))

## 📸 Screenshot


## 📢 To Reviewers
-

orijoon98 commented on plu-domain/src/main/kotlin/com/th/plu/domain/domain/question/Question.kt:
```diff
@@ -3,27 +3,42 @@ package com.th.plu.domain.domain.question
+        private set
+
 
     @Column(name = "question_title", nullable = false, length = 100)
-    var title: String,
+    var title: String = title
+        private set
 
     @Column(name = "question_content", nullable = false, length = 300)
-    var content: String,
-
-    @OneToMany(mappedBy = "question", fetch = FetchType.LAZY, cascade = [CascadeType.ALL], orphanRemoval = true)
-    var answers: List<Answer>
+    var content: String = content
+        private set
 
-) : BaseEntity() {
+    @Column(name = "exposed_at", nullable = false, unique = true)
+    var exposedAt: LocalDateTime = exposedAt
+        private set
```
저 이런식으로 선언하는건 처음보는데, 기존 방식이랑 어떤게 다른건가용?! 궁금합니당 ㅋㅋ
그리고 _id 이렇게 사용하는 부분은 다른 엔티티들에도 한번에 적용해놔도 좋을 것 같아요!

daehwan2da replied:
생성자에 var 을 두면 setter 가 열려서 private 해지게하져고 필드에서 선언해주었어요~

orijoon98 commented on plu-domain/src/main/kotlin/com/th/plu/domain/domain/question/QuestionResult.kt:
```diff
@@ -0,0 +1,11 @@
+
+import java.time.LocalDateTime
+
+data class QuestionResult(
+    val questionId: Long,
+    val title: String,
+    val content: String,
+    val elementType: ElementType,
+    val exposedAt: LocalDateTime,
+)
```
저희 dto 들 네이밍 컨벤션도 맞추면 좋을 것 같아요!
저는 ~RequestDto, ~ResponseDto, ~Dto 이렇게 사용했는데
한번 통일하면 좋을 것 같습니당.

orijoon98 commented on plu-domain/src/main/kotlin/com/th/plu/domain/domain/question/QuestionRetriever.kt:
```diff
@@ -0,0 +1,27 @@
+import java.time.LocalDateTime
+import java.time.YearMonth
+
+@Component
+class QuestionRetriever(
+    private val questionRepository: QuestionRepository,
+) {
+    fun findQuestion(date: LocalDateTime): Question =
+        questionRepository.findByExposedAtOrNull(date) ?: throw InternalServerException(
+            ErrorCode.DATA_NOT_READY_EXCEPTION,
+            "($date) 날짜의 질문데이터가 준비되지 않았습니다. "
+        )
+
+    fun findMyQuestionsMonthly(memberId: Long, yearMonth: YearMonth): List<Question> =
+        questionRepository.findAllByExposedMonthIn(memberId, yearMonth)
+
+    fun findAnsweredYearMonth(memberId: Long): Set<YearMonth> =
+        questionRepository.findAllExposedAtIAnsweredMonth(memberId)
+            .map { YearMonth.of(it.year, it.monthValue) }
+            .toSet() // application 에서 중복 처리중, 500 넘는 warn log 발생시 월별 1건 조회하도록 쿼리 개선 필요!
+}
```
여기도 컨벤션 맞추면 좋을 것 같아요! Member 쪽에는 MemberExplorer 라고 올려둬서..

orijoon98 commented on plu-domain/src/main/kotlin/com/th/plu/domain/domain/question/repository/QuestionRepositoryImpl.kt:
```diff
@@ -1,17 +1,68 @@
+            .where(
+                question.exposedAt.between(
+                    // start (first day of month)
+                    LocalDateTime.of(yearMonth.year, yearMonth.monthValue, 1, 0, 0),
+                    // end (end day of month, 윤년 포함)
+                    LocalDateTime.of(yearMonth.year, yearMonth.monthValue, yearMonth.month.maxLength(), 0, 0),
```
달의 마지막 날의 경우 yyyyMMddHHmmss 를 20240301000000 일 경우에,
between 에 포함될까요? 안전하게 23시 59분 59초 정도로 세팅해줘도 좋을 것 같아요!

daehwan2da replied:
더 안전하게 하기위해
gte(before), lt(later) 로 정의해둘게요

orijoon98 commented on plu-domain/src/main/kotlin/com/th/plu/domain/domain/question/repository/QuestionRepositoryImpl.kt:
```diff
@@ -1,17 +1,68 @@
+    // 쿼리 개선 필요, from 절 sub query 필요 -> JPA 외 다른 tool 사용해야함.
+    // 현재는 작성한 모든 질문의 날짜 조회됨 (full scan, 10000개 썼으면 10000개 read 됨)
+    // 월별로 1개씩 조회되도록 쿼리 개선 필요!
+    override fun findAllExposedAtIAnsweredMonth(memberId: Long): List<LocalDateTime> {
+        return queryFactory
```
typo 한번 확인 부탁드립니다!

orijoon98 commented on plu-api/src/main/kotlin/com/th/plu/api/controller/question/QuestionController.kt:
```diff
@@ -0,0 +1,46 @@
+class QuestionController(
+    private val questionService: QuestionService,
+) {
+    @Auth
+    @Operation(summary = "오늘의 질문")
+    @GetMapping("/api/v1/questions/today")
+    fun getQuestionToday(): ApiResponse<QuestionTodayResponse> = questionService.getQuestionToday().let {
+        ApiResponse.success(toQuestionTodayResponse(it))
+    }
```
오늘의 질문을 답변했을 때, 모두의 답변 보기가 활성화되어야 해서,
오늘의 질문 답변 여부 조회 api 도 필요할 것 같아요!

sookyungg commented on plu-api/src/main/kotlin/com/th/plu/api/controller/question/QuestionController.kt:
```diff
@@ -0,0 +1,46 @@
+    private val questionService: QuestionService,
+) {
+    @Auth
+    @Operation(summary = "오늘의 질문")
```
이 파일의 Operation 부분 명세서와 동일하게 기재해주면 좋을 것 같아요
예를 들면, **[인증] 오늘의 질문 조회** 로요!

daehwan2da committed:
feat : 오늘의 질문

add exposed_at column as unique
get today question api
get question i answered api
get yearMonth i answered api

daehwan2da committed:
implement answer question (#26)
